lib/Transforms/InlineParsers.cpp

```diff
@@ -0,0 +1,252 @@
+            // The transition to subparser's start has been cloned to the "pre" state.
+            // We just want the original op removed, which is handled by InliningUtils.
+        }
+
+        // Update names and operation positions after inlining.
+        void processInlinedBlocks(
+            llvm::iterator_range<mlir::Region::iterator> inlinedBlocks) override {
+            for (mlir::Block &block : inlinedBlocks) {
+                size_t index = 0;
+                for (mlir::Operation &op : llvm::make_early_inc_range(block)) {
+                    proccessInlinedOp(index, &op);
+                    index++;
+                }
+            }
+        }
+
+        // Move inlined operation `op` to its desired position.
+        void proccessInlinedOp(size_t index, mlir::Operation *op) {
+            // Remmap parser locals if needed.
+            if (declMapping[index]) {
+                rewriter.replaceOp(op, declMapping[index]);
+                return;
+            }
+
+            assert((!mlir::isa<P4HIR::ConstOp, P4HIR::VariableOp, P4HIR::InstantiateOp>(op)) &&
+                   "Should be handled by caller");
+
+            P4HIR::ParserOp parser = info.getSplitOp()->getParentOfType<P4HIR::ParserOp>();
+            op->walk([&](mlir::Operation *o) { updateNamesAndRefs(parser, prefix, o); });
+
+            P4HIR::ParserStateOp postState = info.getPostState();
+            mlir::Block *preStateBB = info.getPreState().getBlock();
+
+            if (auto stateOp = mlir::dyn_cast<P4HIR::ParserStateOp>(op)) {
+                rewriter.moveOpBefore(stateOp, postState);
+
+                // Make the subparser's accept transition to the "post" state.
+                if (stateOp.isAccept()) {
+                    auto acceptOp = mlir::cast<P4HIR::ParserAcceptOp>(stateOp.getNextTransition());
+                    rewriter.setInsertionPoint(acceptOp);
+                    rewriter.create<P4HIR::ParserTransitionOp>(acceptOp.getLoc(),
+                                                               postState.getSymbolRef());
+                    rewriter.eraseOp(acceptOp);
+                }
+            } else if (mlir::isa<P4HIR::ParserTransitionOp>(op)) {
+                // Make the subparsers's start transition the terminator of the init block.
+                // We need to clone instead of move because otherwise the inlined block will
+                // be left without a terminator and assert.
+                rewriter.setInsertionPointToEnd(preStateBB);
+                rewriter.clone(*op);
+            } else {
+                // Move other operations (e.g. reads) to the init block.
+                rewriter.moveOpBefore(op, preStateBB, preStateBB->end());
+            }
+        }
+    };
+};
+
+}  // namespace
+
+void InlineParsersPass::runOnOperation() {
+    mlir::IRRewriter rewriter(&getContext());
+
+    // P4 guarantees that subparser definitions come before their use so inlining them
```

Comment:
This is true for P4 input source. But technically since we're referring things via symbols and symbol tables this might be not true anymore and this is not enforced in P4HIR. I would explicitly have a worklist and iterate until it is not empty